common/flatpak-dir.c

```diff
@@ -14628,3 +14628,345 @@ flatpak_dir_delete_mirror_refs (FlatpakDir    *self,
 
   return TRUE;
 }
+
+static gboolean
+maybe_get_metakey_and_origin (FlatpakDir  *dir,
+                              const char  *ref,
+                              GHashTable  *metadata_injection,
+                              GKeyFile   **out_metakey,
+                              char       **out_origin)
+{
+  g_autoptr(GKeyFile) metakey = NULL;
+  g_autofree char *origin = NULL;
+
+  origin = flatpak_dir_get_origin (dir, ref, NULL, NULL);
+  if (origin == NULL)
+    return FALSE;
+
+  if (metadata_injection != NULL)
+    metakey = g_hash_table_lookup (metadata_injection, ref);
+  if (metakey != NULL)
+    g_key_file_ref (metakey);
+  else
+    {
+      g_autoptr(FlatpakDeploy) deploy = flatpak_dir_load_deployed (dir, ref, NULL, NULL, NULL);
+      if (deploy == NULL)
+        return FALSE;
+      metakey = flatpak_deploy_get_metadata (deploy);
+    }
+
+  if (out_metakey)
+    *out_metakey = g_steal_pointer (&metakey);
+  if (out_origin)
+    *out_origin = g_steal_pointer (&origin);
+
+  return TRUE;
+}
+
+static void
+find_used_related_refs (FlatpakDir *dir,
+                        FlatpakDir *system_dir, /* nullable */
+                        GHashTable *used_refs,
+                        GHashTable *metadata_injection,
+                        const char *ref,
+                        GKeyFile   *metakey,
+                        const char *origin)
+{
+  g_autoptr(GPtrArray) related = NULL;
+  int i;
+
+  if (system_dir == NULL)
+    g_hash_table_add (used_refs, g_strdup (ref));
+
+  /* If @system_dir is non-NULL, that means @ref exists in @dir but we should
+   * look in @system_dir for related things */
+  if (system_dir != NULL)
+    related = flatpak_dir_find_local_related_for_metadata (system_dir, ref, origin, metakey, NULL, NULL);
+  else
+    related = flatpak_dir_find_local_related_for_metadata (dir, ref, origin, metakey, NULL, NULL);
+
+  if (related == NULL)
+    return;
+
+  for (i = 0; i < related->len; i++)
+    {
+      FlatpakRelated *rel = g_ptr_array_index (related, i);
+
+      /* Check if this related ref is present in @dir, which implies the one
+       * in @system_dir is NOT the one being used. */
+      if (system_dir != NULL)
+        {
+          g_autoptr(FlatpakDeploy) user_related_deploy = flatpak_dir_load_deployed (dir, rel->ref, NULL, NULL, NULL);
+          if (user_related_deploy != NULL)
+            continue;
+        }
+
+      if (!rel->auto_prune && !g_hash_table_contains (used_refs, rel->ref))
```

Comment:
I think we should check this before doing the above system-dir check as this is  a quick check to avoid doing work.

I.e. start the loop with `if (auto_prune || used_refs.contains(ref)) continue`